intelmq/bots/outputs/smtp_batch/output.py

```diff
@@ -283,31 +327,40 @@ def prepare_mails(self):
             dict_writer.writerow(dict(zip(ordered_fieldnames, ordered_fieldnames)))
             dict_writer.writerows(rows_output)
 
-            email_to = str(mail_record[len(self.key):], encoding="utf-8")
             count = len(rows_output)
-            if not count:
-                path = None
-            else:
-                filename = f'{time.strftime("%y%m%d")}_{count}_events'
-                path = NamedTemporaryFile().name
-
-                with zipfile.ZipFile(path, mode='w', compression=zipfile.ZIP_DEFLATED) as zf:
-                    try:
-                        zf.writestr(filename + ".csv", output.getvalue())
-                    except Exception:
-                        self.logger.error("Error: Cannot zip mail: %r", mail_record)
-                        continue
+            if not count or count == 0:
+                # send no mail if no events are present
+                continue
+
+            # collect all data which must be the same for all events of the
+            # bucket and thus can be used for templating
+            template_data = {
+                k.replace(".", "_"): lines[0][k]
+                for k in ["source.abuse_contact"] + self.additional_grouping_keys
+                if k in rows_output[0]
+            }
+
+            email_to = template_data["source_abuse_contact"]
+            filename = f'{time.strftime("%y%m%d")}_{count}_events'
+            path = NamedTemporaryFile().name
+
+            with zipfile.ZipFile(path, mode='w', compression=zipfile.ZIP_DEFLATED) as zf:
+                try:
+                    zf.writestr(filename + ".csv", output.getvalue())
+                except Exception:
+                    self.logger.error("Error: Cannot zip mail: %r", mail_record)
+                    continue
 
-                if email_to in self.alternative_mail:
-                    print(f"Alternative: instead of {email_to} we use {self.alternative_mail[email_to]}")
-                    email_to = self.alternative_mail[email_to]
+            if email_to in self.alternative_mail:
+                print(f"Alternative: instead of {email_to} we use {self.alternative_mail[email_to]}")
+                email_to = self.alternative_mail[email_to]
 
             mail = Mail(mail_record, email_to, path, count)
-            self.build_mail(mail, send=False)
-            if count:
-                yield mail
+            # build_mail only used to output metadata of the mail -> send=False -> return None
+            self.build_mail(mail, send=False, template_data=template_data)
+            yield (mail, template_data if jinja2 and self.templating and any(self.templating.values()) else {})
 
-    def build_mail(self, mail, send=False, override_to=None):
+    def build_mail(self, mail, send=False, override_to=None, template_data:dict[str, Any]={}):
```

Comment:
Integrate `template_data` into the `Mail` dataclass instead of passing tuples.

Add annotations in the new code for the non-trivial objects (they did not exist when I wrote it originally), ex, `def build_mail(self, mail: Mail,`